// @clerk/shared/dist/runtime/utils/index.js
const require_runtimeEnvironment = require('../runtimeEnvironment-MfBG-86q.js');
const require_handleValueOrFn = require('../handleValueOrFn-KR1byaCL.js');
const require_noop = require('../noop-Bi5xtB31.js');
const require_createDeferredPromise = require('../createDeferredPromise-BgvUcqYv.js');
const require_utils = require('../utils-DJKYBvVp.js');
const require_instance = require('../instance-CDcvbQik.js');

exports.allSettled = require_utils.allSettled;
exports.createDeferredPromise = require_createDeferredPromise.createDeferredPromise;
exports.fastDeepMergeAndKeep = require_utils.fastDeepMergeAndKeep;
exports.fastDeepMergeAndReplace = require_utils.fastDeepMergeAndReplace;
exports.handleValueOrFn = require_handleValueOrFn.handleValueOrFn;
exports.isDevelopmentEnvironment = require_runtimeEnvironment.isDevelopmentEnvironment;
exports.isProductionEnvironment = require_runtimeEnvironment.isProductionEnvironment;
exports.isStaging = require_instance.isStaging;
exports.isTestEnvironment = require_runtimeEnvironment.isTestEnvironment;
exports.logErrorInDevMode = require_utils.logErrorInDevMode;
exports.noop = require_noop.noop;
