// @clerk/shared/dist/runtime/index.js
const require_constants = require('./constants-B8tEdUlH.js');
const require_isomorphicAtob = require('./isomorphicAtob-Chnw1Md3.js');
const require_isomorphicBtoa = require('./isomorphicBtoa-oWNvIgSP.js');
const require_keys = require('./keys-wr08qE7Y.js');
const require_apiUrlFromPublishableKey = require('./apiUrlFromPublishableKey-CY46bNyR.js');
const require_browser = require('./browser-Ch5Nzweq.js');
const require_color = require('./color-BHugulfx.js');
const require_date = require('./date-DNGgJHFN.js');
const require_runtimeEnvironment = require('./runtimeEnvironment-MfBG-86q.js');
const require_deprecated = require('./deprecated-B-fX5ZX6.js');
const require_deriveState = require('./deriveState-Bsd1S6r4.js');
const require_devBrowser = require('./devBrowser-DGSLF_56.js');
const require_error = require('./error-DIeXY7Ki.js');
const require_file = require('./file-CLbjMOtb.js');
const require_getEnvVariable = require('./getEnvVariable-BBmyrE42.js');
const require_handleValueOrFn = require('./handleValueOrFn-KR1byaCL.js');
const require_noop = require('./noop-Bi5xtB31.js');
const require_createDeferredPromise = require('./createDeferredPromise-BgvUcqYv.js');
const require_utils = require('./utils-DJKYBvVp.js');
const require_instance = require('./instance-CDcvbQik.js');
require('./retry-rhkPbMCS.js');
const require_loadScript = require('./loadScript-D8zceEj8.js');
const require_proxy = require('./proxy-B_Yui2Mf.js');
const require_url = require('./url-Clb-n8zC.js');
const require_versionSelector = require('./versionSelector-D5G-cUjs.js');
const require_loadClerkJsScript = require('./loadClerkJsScript-ClPfeAMO.js');
const require_localStorageBroadcastChannel = require('./localStorageBroadcastChannel-Bgbc3aUe.js');
const require_workerTimers = require('./workerTimers-0T9WnoJq.js');
const require_poller = require('./poller-CVTUNwuX.js');
const require_underscore = require('./underscore-C7r5Yoz7.js');
const require_object = require('./object-DrKQtwRq.js');
const require_logger = require('./logger-Dv_S1aDO.js');
require('./pathToRegexp-DNM_UlMK.js');
const require_pathMatcher = require('./pathMatcher-Co1RhryO.js');
const require_netlifyCacheHandler = require('./netlifyCacheHandler-CGV6jGfB.js');

exports.CLERK_NETLIFY_CACHE_BUST_PARAM = require_netlifyCacheHandler.CLERK_NETLIFY_CACHE_BUST_PARAM;
exports.CURRENT_DEV_INSTANCE_SUFFIXES = require_constants.CURRENT_DEV_INSTANCE_SUFFIXES;
exports.ClerkAPIError = require_error.ClerkAPIError;
exports.ClerkAPIResponseError = require_error.ClerkAPIResponseError;
exports.ClerkError = require_error.ClerkError;
exports.ClerkRuntimeError = require_error.ClerkRuntimeError;
exports.ClerkWebAuthnError = require_error.ClerkWebAuthnError;
exports.DEV_BROWSER_JWT_KEY = require_devBrowser.DEV_BROWSER_JWT_KEY;
exports.DEV_OR_STAGING_SUFFIXES = require_constants.DEV_OR_STAGING_SUFFIXES;
exports.EmailLinkError = require_error.EmailLinkError;
exports.EmailLinkErrorCode = require_error.EmailLinkErrorCode;
exports.EmailLinkErrorCodeStatus = require_error.EmailLinkErrorCodeStatus;
exports.LEGACY_DEV_INSTANCE_SUFFIXES = require_constants.LEGACY_DEV_INSTANCE_SUFFIXES;
exports.LOCAL_API_URL = require_constants.LOCAL_API_URL;
exports.LOCAL_ENV_SUFFIXES = require_constants.LOCAL_ENV_SUFFIXES;
exports.LocalStorageBroadcastChannel = require_localStorageBroadcastChannel.LocalStorageBroadcastChannel;
exports.MissingExpiredTokenError = require_error.MissingExpiredTokenError;
exports.PROD_API_URL = require_constants.PROD_API_URL;
exports.Poller = require_poller.Poller;
exports.STAGING_API_URL = require_constants.STAGING_API_URL;
exports.STAGING_ENV_SUFFIXES = require_constants.STAGING_ENV_SUFFIXES;
exports.addClerkPrefix = require_url.addClerkPrefix;
exports.addYears = require_date.addYears;
exports.allSettled = require_utils.allSettled;
exports.apiUrlFromPublishableKey = require_apiUrlFromPublishableKey.apiUrlFromPublishableKey;
exports.applyFunctionToObj = require_object.applyFunctionToObj;
exports.buildClerkJsScriptAttributes = require_loadClerkJsScript.buildClerkJsScriptAttributes;
exports.buildErrorThrower = require_error.buildErrorThrower;
exports.buildPublishableKey = require_keys.buildPublishableKey;
exports.camelToSnake = require_underscore.camelToSnake;
exports.cleanDoubleSlashes = require_url.cleanDoubleSlashes;
exports.clerkJsScriptUrl = require_loadClerkJsScript.clerkJsScriptUrl;
exports.colorToSameTypeString = require_color.colorToSameTypeString;
exports.createClerkGlobalHookError = require_error.createClerkGlobalHookError;
exports.createDeferredPromise = require_createDeferredPromise.createDeferredPromise;
exports.createDevOrStagingUrlCache = require_keys.createDevOrStagingUrlCache;
exports.createPathMatcher = require_pathMatcher.createPathMatcher;
exports.createWorkerTimers = require_workerTimers.createWorkerTimers;
exports.dateTo12HourTime = require_date.dateTo12HourTime;
exports.deepCamelToSnake = require_underscore.deepCamelToSnake;
exports.deepSnakeToCamel = require_underscore.deepSnakeToCamel;
exports.deprecated = require_deprecated.deprecated;
exports.deprecatedObjectProperty = require_deprecated.deprecatedObjectProperty;
exports.deprecatedProperty = require_deprecated.deprecatedProperty;
exports.deriveState = require_deriveState.deriveState;
exports.differenceInCalendarDays = require_date.differenceInCalendarDays;
exports.errorToJSON = require_error.errorToJSON;
exports.extension = require_file.extension;
exports.extractDevBrowserJWTFromURL = require_devBrowser.extractDevBrowserJWTFromURL;
exports.fastDeepMergeAndKeep = require_utils.fastDeepMergeAndKeep;
exports.fastDeepMergeAndReplace = require_utils.fastDeepMergeAndReplace;
exports.filterProps = require_object.filterProps;
exports.formatRelative = require_date.formatRelative;
exports.getClerkJsMajorVersionOrTag = require_url.getClerkJsMajorVersionOrTag;
exports.getCookieSuffix = require_keys.getCookieSuffix;
exports.getEnvVariable = require_getEnvVariable.getEnvVariable;
exports.getNonUndefinedValues = require_underscore.getNonUndefinedValues;
exports.getScriptUrl = require_url.getScriptUrl;
exports.getSuffixedCookieName = require_keys.getSuffixedCookieName;
exports.handleNetlifyCacheInDevInstance = require_netlifyCacheHandler.handleNetlifyCacheInDevInstance;
exports.handleValueOrFn = require_handleValueOrFn.handleValueOrFn;
exports.hasAlpha = require_color.hasAlpha;
exports.hasLeadingSlash = require_url.hasLeadingSlash;
exports.hasTrailingSlash = require_url.hasTrailingSlash;
exports.hexStringToRgbaColor = require_color.hexStringToRgbaColor;
exports.iconImageUrl = require_constants.iconImageUrl;
exports.inBrowser = require_browser.inBrowser;
exports.is4xxError = require_error.is4xxError;
exports.isAbsoluteUrl = require_url.isAbsoluteUrl;
exports.isBrowserOnline = require_browser.isBrowserOnline;
exports.isCaptchaError = require_error.isCaptchaError;
exports.isClerkAPIError = require_error.isClerkAPIError;
exports.isClerkAPIResponseError = require_error.isClerkAPIResponseError;
exports.isClerkError = require_error.isClerkError;
exports.isClerkRuntimeError = require_error.isClerkRuntimeError;
exports.isCurrentDevAccountPortalOrigin = require_url.isCurrentDevAccountPortalOrigin;
exports.isDevelopmentEnvironment = require_runtimeEnvironment.isDevelopmentEnvironment;
exports.isDevelopmentFromPublishableKey = require_keys.isDevelopmentFromPublishableKey;
exports.isDevelopmentFromSecretKey = require_keys.isDevelopmentFromSecretKey;
exports.isEmailLinkError = require_error.isEmailLinkError;
exports.isHSLColor = require_color.isHSLColor;
exports.isHttpOrHttps = require_proxy.isHttpOrHttps;
exports.isIPV4Address = require_underscore.isIPV4Address;
exports.isKnownError = require_error.isKnownError;
exports.isLegacyDevAccountPortalOrigin = require_url.isLegacyDevAccountPortalOrigin;
exports.isMetamaskError = require_error.isMetamaskError;
exports.isNetworkError = require_error.isNetworkError;
exports.isNonEmptyURL = require_url.isNonEmptyURL;
exports.isPasswordCompromisedError = require_error.isPasswordCompromisedError;
exports.isPasswordPwnedError = require_error.isPasswordPwnedError;
exports.isProductionEnvironment = require_runtimeEnvironment.isProductionEnvironment;
exports.isProductionFromPublishableKey = require_keys.isProductionFromPublishableKey;
exports.isProductionFromSecretKey = require_keys.isProductionFromSecretKey;
exports.isProxyUrlRelative = require_proxy.isProxyUrlRelative;
exports.isPublishableKey = require_keys.isPublishableKey;
exports.isRGBColor = require_color.isRGBColor;
exports.isReverificationCancelledError = require_error.isReverificationCancelledError;
exports.isStaging = require_instance.isStaging;
exports.isTestEnvironment = require_runtimeEnvironment.isTestEnvironment;
exports.isTransparent = require_color.isTransparent;
exports.isTruthy = require_underscore.isTruthy;
exports.isUnauthorizedError = require_error.isUnauthorizedError;
exports.isUserLockedError = require_error.isUserLockedError;
exports.isValidBrowser = require_browser.isValidBrowser;
exports.isValidBrowserOnline = require_browser.isValidBrowserOnline;
exports.isValidHexString = require_color.isValidHexString;
exports.isValidHslaString = require_color.isValidHslaString;
exports.isValidProxyUrl = require_proxy.isValidProxyUrl;
exports.isValidRgbaString = require_color.isValidRgbaString;
exports.isomorphicAtob = require_isomorphicAtob.isomorphicAtob;
exports.isomorphicBtoa = require_isomorphicBtoa.isomorphicBtoa;
exports.joinURL = require_url.joinURL;
exports.loadClerkJsScript = require_loadClerkJsScript.loadClerkJsScript;
exports.loadScript = require_loadScript.loadScript;
exports.logErrorInDevMode = require_utils.logErrorInDevMode;
exports.logger = require_logger.logger;
exports.noop = require_noop.noop;
exports.normalizeDate = require_date.normalizeDate;
exports.parseError = require_error.parseError;
exports.parseErrors = require_error.parseErrors;
exports.parsePublishableKey = require_keys.parsePublishableKey;
exports.parseSearchParams = require_url.parseSearchParams;
exports.proxyUrlToAbsoluteURL = require_proxy.proxyUrlToAbsoluteURL;
exports.readJSONFile = require_file.readJSONFile;
exports.removeUndefined = require_object.removeUndefined;
exports.setClerkJsLoadingErrorPackageName = require_loadClerkJsScript.setClerkJsLoadingErrorPackageName;
exports.setDevBrowserJWTInURL = require_devBrowser.setDevBrowserJWTInURL;
exports.snakeToCamel = require_underscore.snakeToCamel;
exports.stringToHslaColor = require_color.stringToHslaColor;
exports.stringToSameTypeColor = require_color.stringToSameTypeColor;
exports.stripScheme = require_url.stripScheme;
exports.titleize = require_underscore.titleize;
exports.toSentence = require_underscore.toSentence;
exports.userAgentIsRobot = require_browser.userAgentIsRobot;
exports.versionSelector = require_versionSelector.versionSelector;
exports.withLeadingSlash = require_url.withLeadingSlash;
exports.withTrailingSlash = require_url.withTrailingSlash;
exports.without = require_object.without;
exports.withoutLeadingSlash = require_url.withoutLeadingSlash;
exports.withoutTrailingSlash = require_url.withoutTrailingSlash;
